### PR TITLE
Faster random walk

### PR DESCRIPTION
**Pull request by nishaq503**
wrote single-threaded random walk that is faster than old multi-threaded random walk

### CODE REVIEW

File: pyclam/manifold.py [hunk 1]
```diff
@@ -564,14 +578,16 @@ class Manifold:
...
+    With this Cluster-tree and Graph-stack, Manifold provides utilities for
+     rho-nearest neighbors search and k-nearest neighbors search.
     """
 
     def __init__(self, data: Data, metric: Metric, argpoints: Union[Vector, float] = None, **kwargs):
         """ A Manifold needs the data to learn the manifold for, and a distance metric to use while doing so.
 
         :param data: The data to learn. This could be a numpy.ndarray or a numpy.memmap.
-        :param metric: The distance metric to use for the data. Any metric allowed by scipy.spatial.distance is allowed here.
+        :param metric: The distance metric to use for the data.
```
[the remark this answers is not in the document]
Comment (nishaq503) [reply pyclam/manifold.py]: yes

File: requirements.txt [hunk 2]
```diff
@@ -1,62 +1,3 @@
...
-pytz==2019.3
-PyYAML==5.1.2
-querystring-parser==1.2.4
-requests==2.22.0
-scipy==1.3.2
-simplejson==3.16.0
-six==1.13.0
-smmap2==2.0.5
-snowballstemmer==2.0.0
-Sphinx==2.2.1
```
Comment (thoward27) [inline requirements.txt]: without this we can't build docs

File: requirements.txt [hunk 3]
```diff
@@ -28,12 +29,16 @@ Mako==1.1.0
...
+more-itertools==8.2.0
 numpy==1.17.4
 packaging==19.2
 pandas==0.25.3
+pluggy==0.13.1
 protobuf==3.10.0
+py==1.8.1
 Pygments==2.4.2
 pyparsing==2.4.5
+pytest==5.3.5
```
[the remark this answers is not in the document]
Comment (nishaq503) [reply requirements.txt]: removed

File: requirements.txt [hunk 4]
```diff
@@ -28,12 +29,16 @@ Mako==1.1.0
 MarkupSafe==1.1.1
 matplotlib==3.1.2
 mlflow==1.4.0
+more-itertools==8.2.0
 numpy==1.17.4
 packaging==19.2
 pandas==0.25.3
+pluggy==0.13.1
```
[the remark this answers is not in the document]
Comment (nishaq503) [reply requirements.txt]: removed

File: pyclam/manifold.py [hunk 5]
```diff
@@ -487,40 +506,36 @@ def clear_cache(self) -> None:
...
-
-                neighbors = list(cluster.neighbors.keys())
-                cluster = np.random.choice(neighbors, p=[probabilities[cluster][n] for n in neighbors])
-
-        with concurrent.futures.ThreadPoolExecutor() as executor:
-            future_to_cluster = [executor.submit(_walk, c) for c in clusters]
-            [v.result() for v in concurrent.futures.as_completed(future_to_cluster)]
+        walks = [c for c in clusters if c.neighbors]
+        for _ in range(steps):
+            walks = [np.random.choice(a=c.transition_probabilities[0], p=c.transition_probabilities[1])
```
[the remark this answers is not in the document]
Comment (nishaq503) [reply pyclam/manifold.py]: we can do this after I've added layer-transitions in random-walk

File: pyclam/manifold.py [hunk 6]
```diff
@@ -637,7 +653,9 @@ def distance(self, x1: Union[List[int], Data], x2: Union[List[int], Data]) -> np
 
     def find_points(self, point: Data, radius: Radius) -> List[Tuple[int, Radius]]:
         """ Returns all indices of points that are within radius of point. """
-        candidates: List[int] = [p for c in self.find_clusters(point, radius, len(self.graphs)).keys() for p in c.argpoints]
+        candidates: List[int] = [p
+                                 for c in self.find_clusters(point, radius, len(self.graphs))
```
[the remark this answers is not in the document]
Comment (nishaq503) [reply pyclam/manifold.py]: switched

**Pull request merged**